### PR TITLE
Refactor DefaultSolver

### PR DESCRIPTION
**Pull request by AntoniusW**
Some refactoring/polishing inside the solver component, mostly DefaultSolver. Changes are intended to support subclassing of DefaultSolver for optimization tasks/weak constraints.

### CODE REVIEW

File: src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java [hunk 1]
```diff
@@ -124,108 +127,124 @@ private BranchingHeuristic chainFallbackHeuristic(Grounder grounder, WritableAss
...
+	private void setupLoop() {
+		if (!searchState.hasBeenInitialized) {
+			// Initially, get NoGoods from grounder.
+			performanceLog.initialize();
+			getNoGoodsFromGrounderAndIngest();
+			searchState.hasBeenInitialized = true;
+			return;
+		}
+		// We already found one Answer-Set and are requested to find another one.
+		searchState.afterAllAtomsAssigned = false;
```
Comment (madmike200590) [inline src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java:169]: If I read the code correctly, we reach this part of the code on the second (and subsequent) call(s) to tryAdvance (as in the first call, we'll return after initializing search space). This is not really intuitive from the comment here, I think readability would benefit from a somewhat more detailed explanation on _why_ we only get there after finding an answer set in the comment.

**Review by madmike200590 — APPROVED** [review]
This looks really good! The additional methods in the solver improve readability quite a bit. There's one minor comment where I think a comment could be a bit clearer, but as far as I'm concerned, this is ready to be merged.

File: src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java (hunk 1 above)
Comment (AntoniusW) [reply src/main/java/at/ac/tuwien/kr/alpha/solver/DefaultSolver.java:176]: Yes, the code is doing that. I split up the method into two separate ones.

**Comment by AntoniusW** [comment]
@lorenzleutgeb and/or @madmike200590 please have another look.

**Review by madmike200590 — APPROVED** [review]
Looks good to me! I think the newly split version of the solver initialization part enhances readability a lot.

**Pull request merged**